-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40634/34263

ERROR: Build errors found during clang-tidy run.
```
SimCalorimetry/HcalSimAlgos/src/HcalSiPM.cc:93:17: error: expected expression [clang-diagnostic-error]
    for (i = 0; , < 170; i++) {
                ^
SimCalorimetry/HcalSimAlgos/src/HcalSiPM.cc:93:19: error: expected expression [clang-diagnostic-error]
    for (i = 0; , < 170; i++) {
                  ^
Suppressed 532 warnings (531 in non-user code, 1 with check filters).
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:129: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

--- a/CondFormats/HcalObjects/src/HcalSiPMCharacteristics.cc
+++ b/CondFormats/HcalObjects/src/HcalSiPMCharacteristics.cc
@@ -77,6 +77,7 @@ int HcalSiPMCharacteristics::getPixels(int type) const {
 std::vector<float> HcalSiPMCharacteristics::getNonLinearities(int type) const {
   const HcalSiPMCharacteristics::PrecisionItem* item = findByType(type);
   std::vector<float> pars;
+  pars.reserve(3);
   if (item) {
     pars.push_back(item->parLin1_);
     pars.push_back(item->parLin2_);

--- a/SimCalorimetry/HcalSimAlgos/interface/HcalSiPM.h
+++ b/SimCalorimetry/HcalSimAlgos/interface/HcalSiPM.h
@@ -19,17 +19,17 @@ namespace CLHEP {
   class HepRandomEngine;
 }
 
-class HcalSiPM {
+class HcalSiPM final {
 public:
   HcalSiPM(int nCells = 1, double tau = 15.);
 
-  virtual ~HcalSiPM();
+  ~HcalSiPM();
 
-  void resetSiPM() { std::fill(theSiPM.begin(), theSiPM.end(), -999.); }
-  virtual double hitCells(CLHEP::HepRandomEngine*, unsigned int pes, double tempDiff = 0., double photonTime = 0.);
+  void resetSiPM() { std::fill(theSiPM.begin(), theSiPM.end(), -999.f); }
+  double hitCells(CLHEP::HepRandomEngine*, unsigned int pes, double tempDiff = 0., double photonTime = 0.);
 
-  virtual double totalCharge() const { return totalCharge(theLastHitTime); }
-  virtual double totalCharge(double time) const;
+  double totalCharge() const { return totalCharge(theLastHitTime); }
+  double totalCharge(double time) const;
 
   int getNCells() const { return theCellCount; }
   double getTau() const { return theTau; }
@@ -52,11 +52,10 @@ protected:
   unsigned int addCrossTalkCells(CLHEP::HepRandomEngine* engine, unsigned int in_pes);
 
   //numerical random generation from Borel-Tanner distribution
-  double Borel(unsigned int n, double lambda, unsigned int k);
   const cdfpair& BorelCDF(unsigned int k);
 
   unsigned int theCellCount;
-  std::vector<double> theSiPM;
+  std::vector<float> theSiPM;
   double theTau;
   double theTauInv;
   double theCrossTalk;

--- a/SimCalorimetry/HcalSimAlgos/interface/HcalSiPMShape.h
+++ b/SimCalorimetry/HcalSimAlgos/interface/HcalSiPMShape.h
@@ -2,17 +2,24 @@
 #ifndef HcalSimAlgos_HcalSiPMShape_h
 #define HcalSimAlgos_HcalSiPMShape_h
 
+#include "CalibCalorimetry/HcalAlgos/interface/HcalPulseShapes.h"
 #include "SimCalorimetry/CaloSimAlgos/interface/CaloVShape.h"
 #include <vector>
 
-class HcalSiPMShape : public CaloVShape {
+class HcalSiPMShape final : public CaloVShape {
 public:
   HcalSiPMShape(unsigned int signalShape = 206);
   HcalSiPMShape(const HcalSiPMShape& other);
 
   ~HcalSiPMShape() override {}
 
-  double operator()(double time) const override;
+  int nBins() const { return nBins_; }
+  double operator[](int i) const { return nt_[i]; }
+
+  double operator()(double time) const override {
+    int jtime(time * HcalPulseShapes::invDeltaTSiPM_ + 0.5);
+    return (jtime >= 0 && jtime < nBins_) ? nt_[jtime] : 0;
+  }
 
   double timeToRise() const override { return 0.0; }
 

--- a/SimCalorimetry/HcalSimAlgos/src/HcalSiPM.cc
+++ b/SimCalorimetry/HcalSimAlgos/src/HcalSiPM.cc
@@ -81,6 +81,7 @@ const HcalSiPM::cdfpair& HcalSiPM::BorelCDF(unsigned int k) {
   // EPSILON determines the min and max # of xtalk cells that can be
   // simulated.
   constexpr double EPSILON = 1e-6;
+  constexpr uint32_t maxCDFsize = 170;  // safe max to avoid factorial to be infinite
   auto it = borelcdfs.find(k);
   if (it == borelcdfs.end()) {
     vector<double> cdf;
@@ -90,7 +91,7 @@ const HcalSiPM::cdfpair& HcalSiPM::BorelCDF(unsigned int k) {
     unsigned int i;
     double sumb = 0.;
     double iFact = 1.;
-    for (i = 0; i < 170; i++) {
+    for (i = 0; i < maxCDFsize; i++) {
       if (i > 0)
         iFact *= double(i);
       sumb += Borel(i, theCrossTalk, k, iFact);
@@ -102,7 +103,7 @@ const HcalSiPM::cdfpair& HcalSiPM::BorelCDF(unsigned int k) {
     unsigned int borelstartn = i;
 
     // calculate cdf[i]  limit to 170 to avoid iFact to become infinite
-    for (++i; i < 170; ++i) {
+    for (++i; i < maxCDFsize; ++i) {
       iFact *= double(i);
       sumb += Borel(i, theCrossTalk, k, iFact);
       cdf.push_back(sumb);

--- a/SimCalorimetry/HcalSimAlgos/src/HcalSiPM.cc
+++ b/SimCalorimetry/HcalSimAlgos/src/HcalSiPM.cc
@@ -90,7 +90,7 @@ const HcalSiPM::cdfpair& HcalSiPM::BorelCDF(unsigned int k) {
     unsigned int i;
     double sumb = 0.;
     double iFact = 1.;
-    for (i = 0; , < 170; i++) {
+    for (i = 0; i < 170; i++) {
       if (i > 0)
         iFact *= double(i);
       sumb += Borel(i, theCrossTalk, k, iFact);

--- a/SimCalorimetry/HcalSimAlgos/src/HcalSiPMHitResponse.cc
+++ b/SimCalorimetry/HcalSimAlgos/src/HcalSiPMHitResponse.cc
@@ -205,59 +205,69 @@ CaloSamples HcalSiPMHitResponse::makeSiPMSignal(DetId const& id,
   unsigned int pe(0);
   double hitPixels(0.), elapsedTime(0.);
 
-  auto& sipmPulseShape(shapeMap[pars.signalShape(id)]);
+  auto const& sipmPulseShape(shapeMap[pars.signalShape(id)]);
 
-  std::list<std::pair<double, double> > pulses;
-  std::list<std::pair<double, double> >::iterator pulse;
-  double timeDiff, pulseBit;
   LogDebug("HcalSiPMHitResponse") << "makeSiPMSignal for " << HcalDetId(id);
 
-  for (unsigned int tbin(0); tbin < photonTimeBins.size(); ++tbin) {
+  int nptb = photonTimeBins.size();
+  double sum[nptb];
+  for (auto i = 0; i < nptb; ++i)
+    sum[i] = 0;
+  for (int tbin(0); tbin < nptb; ++tbin) {
     pe = photonTimeBins[tbin];
+    if (pe <= 0)
+      continue;
     preciseBin = tbin;
     sampleBin = preciseBin / nbins;
-    if (pe > 0) {
-      //skip saturation/recovery and pulse smearing for premix stage 1
-      if (PreMixDigis and HighFidelityPreMix) {
-        signal[sampleBin] += pe;
-        signal.preciseAtMod(preciseBin) += pe;
-        elapsedTime += dt;
-        continue;
-      }
-
-      hitPixels = theSiPM.hitCells(engine, pe, 0., elapsedTime);
-      LogDebug("HcalSiPMHitResponse") << " elapsedTime: " << elapsedTime << " sampleBin: " << sampleBin
-                                      << " preciseBin: " << preciseBin << " pe: " << pe << " hitPixels: " << hitPixels;
-      if (pars.doSiPMSmearing()) {
-        pulses.push_back(std::pair<double, double>(elapsedTime, hitPixels));
-      } else {
-        signal[sampleBin] += hitPixels;
-        signal.preciseAtMod(preciseBin) += 0.6 * hitPixels;
-        if (preciseBin > 0)
-          signal.preciseAtMod(preciseBin - 1) += 0.2 * hitPixels;
-        if (preciseBin < signal.preciseSize() - 1)
-          signal.preciseAtMod(preciseBin + 1) += 0.2 * hitPixels;
-      }
+    //skip saturation/recovery and pulse smearing for premix stage 1
+    if (PreMixDigis and HighFidelityPreMix) {
+      signal[sampleBin] += pe;
+      signal.preciseAtMod(preciseBin) += pe;
+      elapsedTime += dt;
+      continue;
     }
 
-    if (pars.doSiPMSmearing()) {
-      pulse = pulses.begin();
-      while (pulse != pulses.end()) {
-        timeDiff = elapsedTime - pulse->first;
-        pulseBit = sipmPulseShape(timeDiff) * pulse->second;
-        LogDebug("HcalSiPMHitResponse") << " pulse t: " << pulse->first << " pulse A: " << pulse->second
-                                        << " timeDiff: " << timeDiff << " pulseBit: " << pulseBit;
-        signal[sampleBin] += pulseBit;
-        signal.preciseAtMod(preciseBin) += pulseBit;
-
-        if (timeDiff > 1 && sipmPulseShape(timeDiff) < 1e-7)
-          pulse = pulses.erase(pulse);
-        else
-          ++pulse;
+    hitPixels = theSiPM.hitCells(engine, pe, 0., elapsedTime);
+    LogDebug("HcalSiPMHitResponse") << " elapsedTime: " << elapsedTime << " sampleBin: " << sampleBin
+                                    << " preciseBin: " << preciseBin << " pe: " << pe << " hitPixels: " << hitPixels;
+    if (!pars.doSiPMSmearing()) {
+      signal[sampleBin] += hitPixels;
+      signal.preciseAtMod(preciseBin) += 0.6 * hitPixels;
+      if (preciseBin > 0)
+        signal.preciseAtMod(preciseBin - 1) += 0.2 * hitPixels;
+      if (preciseBin < signal.preciseSize() - 1)
+        signal.preciseAtMod(preciseBin + 1) += 0.2 * hitPixels;
+    } else {
+      // add "my" smearing to future bins...
+      // this loop can vectorize....
+      for (auto i = tbin; i < nptb; ++i) {
+        auto itdiff = i - tbin;
+        if (itdiff == sipmPulseShape.nBins())
+          break;
+        auto shape = sipmPulseShape[itdiff];
+        auto pulseBit = shape * hitPixels;
+        sum[i] += pulseBit;
+        if (shape < 1.e-7 && itdiff > int(HcalPulseShapes::invDeltaTSiPM_))
+          break;
       }
     }
     elapsedTime += dt;
   }
+  if (pars.doSiPMSmearing())
+    for (auto i = 0; i < nptb; ++i) {
+      auto iSampleBin = i / nbins;
+      signal[iSampleBin] += sum[i];
+      signal.preciseAtMod(i) += sum[i];
+    }
+
+#ifdef EDM_ML_DEBUG
+  LogDebug("HcalSiPMHitResponse") << nbins << ' ' << nptb << ' ' << HcalDetId(id);
+  for (auto i = 0; i < nptb; ++i) {
+    auto iSampleBin = (nbins > 1) ? i / nbins : i;
+    LogDebug("HcalSiPMHitResponse") << i << ' ' << iSampleBin << ' ' << signal[iSampleBin] << ' '
+                                    << signal.preciseAtMod(i);
+  }
+#endif
 
   return signal;
 }

--- a/SimCalorimetry/HcalSimAlgos/src/HcalSiPMHitResponse.cc
+++ b/SimCalorimetry/HcalSimAlgos/src/HcalSiPMHitResponse.cc
@@ -209,7 +209,7 @@ CaloSamples HcalSiPMHitResponse::makeSiPMSignal(DetId const& id,
 
   LogDebug("HcalSiPMHitResponse") << "makeSiPMSignal for " << HcalDetId(id);
 
-  int nptb = photonTimeBins.size();
+  const int nptb = photonTimeBins.size();
   double sum[nptb];
   for (auto i = 0; i < nptb; ++i)
     sum[i] = 0;

--- a/SimCalorimetry/HcalSimAlgos/src/HcalSiPMShape.cc
+++ b/SimCalorimetry/HcalSimAlgos/src/HcalSiPMShape.cc
@@ -10,13 +10,6 @@ HcalSiPMShape::HcalSiPMShape(unsigned int signalShape)
 
 HcalSiPMShape::HcalSiPMShape(const HcalSiPMShape& other) : CaloVShape(other), nBins_(other.nBins_), nt_(other.nt_) {}
 
-double HcalSiPMShape::operator()(double time) const {
-  int jtime(time * HcalPulseShapes::invDeltaTSiPM_ + 0.5);
-  if (jtime >= 0 && jtime < nBins_)
-    return nt_[jtime];
-  return 0.;
-}
-
 void HcalSiPMShape::computeShape(unsigned int signalShape) {
   //grab correct function pointer based on shape
   double (*analyticPulseShape)(double);

